File: src/main/java/com/everchanging/hierarchy/converter/NamesLoopToValidationErrorConverter.java
package com.everchanging.hierarchy.converter;

import com.everchanging.hierarchy.dto.ValidationError;
import org.springframework.core.convert.converter.Converter;
import org.springframework.stereotype.Component;

import java.util.ArrayList;

import static java.lang.String.format;

@Component
public class NamesLoopToValidationErrorConverter implements Converter<ArrayList<String>, ValidationError> {
    @Override
    public ValidationError convert(ArrayList<String> nameLoops) {
        String nameLoopMessage = buildNameLoopMessage(nameLoops);
        return new ValidationError("HierarchyLoop", format("Found loop with the Employees: [%s]", nameLoopMessage));
    }

    private String buildNameLoopMessage(ArrayList<String> nameLoops) {
        if (nameLoops.size() >= 5) {
            return format("%s -> (%d employees) -> %s -> %s",
                    nameLoops.get(0), nameLoops.size()-3, nameLoops.get(nameLoops.size()-2), nameLoops.get(nameLoops.size()-1));
        }
        StringBuilder message = new StringBuilder();
        for (int i = 0; i < nameLoops.size(); i++) {
            String name = nameLoops.get(i);
            message.append(name);
            if (i != nameLoops.size()-1) {
                message.append(" -> ");
            }
        }
        return message.toString();
    }
}

File: src/main/kotlin/com/everchanging/hierarchy/converter/NamesLoopToValidationErrorConverter.kt
package com.everchanging.hierarchy.converter

import com.everchanging.hierarchy.dto.ValidationError
import org.springframework.core.convert.converter.Converter
import org.springframework.stereotype.Component
import java.lang.StringBuilder

@Component
class NamesLoopToValidationErrorConverter : Converter<ArrayList<String>, ValidationError> {

    override fun convert(nameLoops: ArrayList<String>): ValidationError {
        val nameLoopMessage = buildNameLoopMessage(nameLoops)
        return ValidationError("HierarchyLoop", "Found loop with the Employees: [$nameLoopMessage]")
    }

    private fun buildNameLoopMessage(nameLoops: ArrayList<String>): String {
        if (nameLoops.size >= 5) {
            return "${nameLoops[0]} -> (${nameLoops.size - 3} employees) -> ${nameLoops[nameLoops.lastIndex - 1]} -> ${nameLoops[nameLoops.lastIndex]}"
        }
        val message = StringBuilder()
        for ((idx, name) in nameLoops.withIndex()) {
            message.append(name)
            if (idx != nameLoops.lastIndex) {
                message.append(" -> ")
            }
        }
        return message.toString()
    }

}